### ISSUE TITLE
nixos-rebuild in nix-shell produces corrupt grub.cfg

### ISSUE THREAD

**Issue by erictapen** [NixOS/nixpkgs#28598]
## Issue description

`nixos-rebuild` produces a broken `grub.cfg` in nixos-17.03 if executed in a `nix-shell`.

### Steps to reproduce

1) Start NixOS 17.03-1700, preferably in a virtual environment as this makes the `grub.cfg` corrupt.
2) As root: `nix-shell -p vim`
3) Change something that affects Kernel or `initrd` (e.g. add some unused kernel module like `usb_storage` to `boot.initrd.availableKernelModules`) and rebuild with `nixos-rebuild switch`.

 Inspect `/boot/grub/grub.cfg`. It now contains paths to linux kernel of the form
```
menuentry "NixOS - Default" {
search --set=drive1 --label nixos
search --set=drive2 --label nixos
  linux ($drive2)//y10bb973l89wl5n3gwx8zl0ml4h24hg6-linux-4.4.45/bzImage systemConfig=/nix/store/vhsfgs800nrgcacx1kz3h0mqhznlrr8i-nixos-system-nixos-16.09.1623.74f24c7 init=/nix/store/vhsfgs800nrgcacx1kz3h0mqhznlrr8i-nixos-system-nixos-16.09.1623.74f24c7/init loglevel=4
  initrd ($drive2)//jlj36qysn6s1jxn5pwiz3lvnk3m0pmhf-initrd/initrd
}
```

The path to `linux` and `initrd` is missing a `nix/store` in front. This applies to all GRUB menu entries, not just the default one.

If rebuilt outside of `nix-shell`, the `grub.cfg` is normal again.

## Technical details

* System: 17.03.1700.51a83266d1 (Gorilla)
* Nix version: 1.11.13
* Nixpkgs version: "17.03.1700.51a83266d1"
* Sandboxing enabled: false

**Comment by FRidh** [comment]
I think we should check for `$IN_NIX_SHELL`.

**Comment by dezgeg** [comment]
Dupe of https://github.com/NixOS/nixpkgs/issues/14999 and fix is https://github.com/NixOS/nixpkgs/pull/27367 (I think).

**Comment by fpletz** [comment]
Merged #27367.
